signac/core/jsoncollection.py

```diff
@@ -6,59 +6,97 @@
 
-    def __init__(self, filename=None, write_concern=False, **kwargs):
-        self._filename = os.path.realpath(filename) if filename is not None else None
-        self._write_concern = write_concern
+    def __init__(self, filename=None, data=None, write_concern=False, no_sync=False, **kwargs):
+        kwargs['data'] = data
         super().__init__(**kwargs)
         if (filename is None) == (self._parent is None):
             raise ValueError(
                 "Illegal argument combination, one of the two arguments, "
                 "parent or filename must be None, but not both.")
+        self.backend_kwargs['filename'] = None if filename is None else os.path.realpath(filename)
+        self.backend_kwargs['write_concern'] = write_concern
+        self.backend_kwargs['backend'] = self.backend
+        self._id = str(uuid.uuid5(get_namespace(type(self).__name__),
+                       self.backend_kwargs['filename'])) if filename else None
+        if not no_sync and data is not None:
+            self.sync()
 
     def _load(self):
         """Load the data from a JSON-file."""
         try:
-            with open(self._filename, 'rb') as file:
+            with open(self.backend_kwargs['filename'], 'rb') as file:
                 blob = file.read()
                 return json.loads(blob)
         except IOError as error:
             if error.errno == errno.ENOENT:
                 return None
 
-    def _sync(self):
+    def _sync(self, data=None):
         """Write the data to json file."""
-        data = self.to_base()
+        if data is None:
+            data = self.to_base()
+        _filename = self.backend_kwargs['filename']
         # Serialize data:
         blob = json.dumps(data).encode()
         # When write_concern flag is set, we write the data into dummy file and then
         # replace that file with original file.
-        if self._write_concern:
-            dirname, filename = os.path.split(self._filename)
+        if self.backend_kwargs['write_concern']:
+            dirname, filename = os.path.split(_filename)
             fn_tmp = os.path.join(dirname, '._{uid}_{fn}'.format(
                 uid=uuid.uuid4(), fn=filename))
             with open(fn_tmp, 'wb') as tmpfile:
                 tmpfile.write(blob)
-            os.replace(fn_tmp, self._filename)
+            os.replace(fn_tmp, _filename)
         else:
-            with open(self._filename, 'wb') as file:
+            with open(_filename, 'wb') as file:
                 file.write(blob)
 
+    @classmethod
+    def _write_to_cache(cls, id, data, cache):
+        # Serialize data
+        blob = json.dumps(data)
+        cache[id] = blob
+
+    def write_to_cache(self, data=None):
+        data = self.to_base() if data is None else data
+        self._write_to_cache(self._id, data, self._cache)
+
+    @classmethod
+    def _read_from_cache(cls, id, cache):
+        try:
+            data = cache[id]
+        except Exception:
```

Comment:
Never, never catch a bare `Exception`. This will also catch tons of things that you wouldn't expect (like keyboard commands to kill a process like Ctrl+C!). You're looking for an element of a container by key, so catch `KeyError`.

```suggestion
        except KeyError:
```

If it's possible to raise any other exceptions (Redis errors?), then you can catch those as well by putting them in a tuple:

```suggestion
        except (KeyError, SomeRedisException):
```